src/evotorch/core.py

```diff
@@ -2377,7 +2392,7 @@ def sample_and_compute_gradients(
 
             if torch.device(self.device) != torch.device("cpu"):
                 # If the main device of this problem instance is not CPU, then we move the tensors to the main device.
-                result = cast_tensors_in_container(mapresult, device=device)
+                result = cast_tensors_in_container(result, device=self.device)
```

Comment:
I can't work out where `mapresult` was being defined previously, as there's no change I can see that is removing its definition... this is concerning me a little but I am probably just missing it?

Reply:
Thanks for the comment and the concern is valid: the old version of this line was the remnant of a half-finished refactoring. This change is meant to fix this.